// nwb/lib/createServeReactAppConfig.js
'use strict';

exports.__esModule = true;
exports.default = createServeReactAppConfig;

var _path = require('path');

var _path2 = _interopRequireDefault(_path);

var _glob = require('glob');

var _glob2 = _interopRequireDefault(_glob);

var _webpackMerge = require('webpack-merge');

var _webpackMerge2 = _interopRequireDefault(_webpackMerge);

var _appConfig = require('./appConfig');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function createServeReactAppConfig(args, overrides) {
  var entry = _path2.default.resolve(args._[1] || 'src/index.js');
  var dist = _path2.default.resolve(args._[2] || 'dist');

  var config = {
    babel: {
      commonJSInterop: true,
      presets: ['react', 'react-hmre']
    },
    entry: [_path2.default.resolve(entry)],
    output: {
      path: _path2.default.resolve(dist),
      filename: 'app.js',
      publicPath: '/'
    },
    plugins: {
      html: (0, _appConfig.getDefaultHTMLConfig)()
    }
  };

  if (_glob2.default.sync('public/').length !== 0) {
    config.plugins.copy = [{ from: _path2.default.resolve('public'), to: dist, ignore: '.gitkeep' }];
  }

  return (0, _webpackMerge2.default)(config, overrides);
}
module.exports = exports['default'];
