// nwb/lib/createServerWebpackConfig.js
'use strict';

exports.__esModule = true;

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

exports.default = createServerWebpackConfig;

var _createWebpackConfig = require('./createWebpackConfig');

var _createWebpackConfig2 = _interopRequireDefault(_createWebpackConfig);

var _getPluginConfig = require('./getPluginConfig');

var _getPluginConfig2 = _interopRequireDefault(_getPluginConfig);

var _getUserConfig = require('./getUserConfig');

var _getUserConfig2 = _interopRequireDefault(_getUserConfig);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

/**
 * Creates the final Webpack config for serving a web app with hot reloading,
 * using build and user configuration.
 */
function createServerWebpackConfig(args, buildConfig) {
  var userConfig = (0, _getUserConfig2.default)(args);
  var pluginConfig = (0, _getPluginConfig2.default)({ args: args });

  var entry = buildConfig.entry,
      output = buildConfig.output,
      _buildConfig$plugins = buildConfig.plugins,
      plugins = _buildConfig$plugins === undefined ? {} : _buildConfig$plugins,
      otherBuildConfig = _objectWithoutProperties(buildConfig, ['entry', 'output', 'plugins']);

  var hotMiddlewareOptions = args.reload ? '?reload=true' : '';

  if (args['auto-install'] || args.install) {
    plugins.autoInstall = true;
  }

  return (0, _createWebpackConfig2.default)(_extends({
    server: true,
    devtool: 'cheap-module-source-map',
    entry: [
    // Polyfill EventSource for IE, as webpack-hot-middleware/client uses it
    require.resolve('eventsource-polyfill'), require.resolve('webpack-hot-middleware/client') + hotMiddlewareOptions].concat(entry),
    output: output,
    plugins: plugins
  }, otherBuildConfig), pluginConfig, userConfig);
}
module.exports = exports['default'];
