// nwb/lib/commands/serve-react-demo.js
'use strict';

exports.__esModule = true;
exports.default = serveReactDemo;

var _path = require('path');

var _path2 = _interopRequireDefault(_path);

var _webpackServer = require('../webpackServer');

var _webpackServer2 = _interopRequireDefault(_webpackServer);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * Serve a React demo app from demo/src/index.js.
 */
function serveReactDemo(args, cb) {
  var pkg = require(_path2.default.resolve('package.json'));

  (0, _webpackServer2.default)(args, {
    babel: {
      commonJSInterop: true,
      presets: ['react', 'react-hmre']
    },
    entry: [_path2.default.resolve('demo/src/index.js')],
    output: {
      filename: 'demo.js',
      // This doesn't really matter, as files will be served from memory
      path: process.cwd(),
      publicPath: '/'
    },
    plugins: {
      html: {
        mountId: 'demo',
        title: pkg.name + ' ' + pkg.version + ' Demo'
      }
    }
  }, cb);
}
module.exports = exports['default'];
