// nwb/lib/commands/serve-inferno.js
'use strict';

exports.__esModule = true;
exports.default = serveInferno;

var _path = require('path');

var _path2 = _interopRequireDefault(_path);

var _resolve = require('resolve');

var _resolve2 = _interopRequireDefault(_resolve);

var _runSeries = require('run-series');

var _runSeries2 = _interopRequireDefault(_runSeries);

var _errors = require('../errors');

var _webpackServer = require('../webpackServer');

var _webpackServer2 = _interopRequireDefault(_webpackServer);

var _utils = require('../utils');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

// Using a config function as we may need to resolve the path to Inferno, which
// we may also have to install first.
function buildConfig(args) {
  var entry = args._[1];
  var mountId = args['mount-id'] || 'app';

  var config = {
    babel: {
      commonJSInterop: true,
      presets: ['inferno'],
      stage: 0
    },
    output: {
      filename: 'app.js',
      path: process.cwd(),
      publicPath: '/'
    },
    plugins: {
      html: {
        mountId: mountId,
        title: args.title || 'Inferno App'
      }
    },
    resolve: {
      alias: {
        'react': 'inferno-compat',
        'react-dom': 'inferno-compat'
      }
    }
  };

  if (args.force === true) {
    config.entry = [_path2.default.resolve(entry)];
  } else {
    // Use a render shim module which supports quick prototyping
    config.entry = [require.resolve('../infernoRunEntry')];
    config.plugins.define = { NWB_INFERNO_RUN_MOUNT_ID: JSON.stringify(mountId) };
    // Allow the render shim module to resolve Inferno from the cwd
    config.resolve.alias['inferno'] = _path2.default.dirname(_resolve2.default.sync('inferno/package.json', { basedir: process.cwd() }));
    // Allow the render shim module to import the provided entry module
    config.resolve.alias['nwb-inferno-run-entry'] = _path2.default.resolve(entry);
  }

  if (args.polyfill === false || args.polyfills === false) {
    config.polyfill = false;
  }

  return config;
}

/**
 * Build a standalone Inferno entry module, component or VNode.
 */
function serveInferno(args, cb) {
  if (args._.length === 1) {
    return cb(new _errors.UserError('An entry module must be specified.'));
  }

  (0, _runSeries2.default)([function (cb) {
    return (0, _utils.install)(['inferno', 'inferno-component', 'inferno-compat'], { args: args, check: true }, cb);
  }, function (cb) {
    return (0, _webpackServer2.default)(args, buildConfig, cb);
  }], cb);
}
module.exports = exports['default'];
